BlueprintUI/Sources/Layout/Decorate.swift

```diff
@@ -0,0 +1,236 @@
+/// │ ┗━━━━━━━━━━━━━━━┛ │ ▼     ┗━━━━━━━━━━━━━━━┛   ▼
+/// └───────────────────┘
+///   ◀───────────────▶         ◀───────────────▶
+/// ```
+public struct Decorate : ProxyElement {
+    
+    /// The element which provides the sizing and measurement.
+    /// The sizing and position of the `Decorate` element is determined
+    /// by this element.
+    public var wrapped : Element
+    
+    /// The element which is used to draw the decoration.
+    /// It does not affect sizing or positioning.
+    public var decoration : Element
+    
+    /// Where the decoration should be positioned in the z-axis: Above or below the wrapped element.
+    public var layering : Layering
+    
+    /// How the `decoration` should be positioned in respect to the `wrapped` element.
+    public var position : Position
+    
+    /// Creates a new instance with the provided overflow, background, and wrapped element.
+    public init(
+        layering : Layering,
+        position: Position,
+        wrapping: () -> Element,
+        decoration: () -> Element
+    ) {
+        self.layering = layering
+        
+        self.wrapped = wrapping()
+        
+        self.position = position
+        self.decoration = decoration()
+    }
+    
+    // MARK: ProxyElement
+    
+    public var elementRepresentation: Element {
+        EnvironmentReader { environment in
+            LayoutWriter { context, layout in
+                
+                let contentSize = self.wrapped.content.measure(
+                    in: context.size,
+                    environment: environment
+                )
+                
+                let contentFrame = CGRect(origin: .zero, size: contentSize)
+                
+                let decorationFrame = self.position.frame(
+                    with: contentFrame,
+                    decoration: self.decoration,
+                    environment: environment
+                )
+                
+                layout.sizing = .fixed(contentSize)
+                
+                switch self.layering {
+                case .above:
+                    layout.add(with: contentFrame, child: self.wrapped)
+                    layout.add(with: decorationFrame, child: self.decoration)
+                    
+                case .below:
+                    layout.add(with: decorationFrame, child: self.decoration)
+                    layout.add(with: contentFrame, child: self.wrapped)
+                }
+            }
+        }
+    }
+}
+
+
+extension Decorate {
+    
+    /// If the decoration should be positioned above or below the content element.
+    public enum Layering : Equatable {
+        
+        /// The decoration is displayed above the content element.
+        case above
+        
+        /// The decoration is displayed below the content element.
+        case below
+    }
+    
+    /// What corner the decoration element should be positioned in.
+    public enum Corner : Equatable {
```

Comment:
I'm not sure this is worth it to lose the custom frame case – providing an "escape hatch" for people still seems valuable so we're not on the hook for implementing every needed case within this type.

Reply:
What about replacing Corner with Alignment; seems about the same but added functionality for centering on a given side?

Reply:
Actually, I think to do this "right" it's gonna take unifying some types between alignment guides, the aligned type, and this. Let's go ahead and merge this as is, and then I can follow up with that in another PR

Reply:
Hmm, this is really the main sticking point for me in this PR. If we use Alignment it's almost exactly the same as SwiftUI's secondary views, but if we invent new types for this it moves us further away.

I guess I would be OK merging as-is as long as we follow up before we do another release? Alternatively, I could make a PR onto this branch with suggested changes, seeing as how I've worked with alignments already, I have a pretty good idea of the changes involved.

> What about replacing Corner with Alignment; seems about the same but added functionality for centering on a given side?

It's not. The static cases have similar names, but alignments are opaque — resolving them gives you an x/y position in the coordinate space of the parent.

> providing an "escape hatch" for people still seems valuable

I think the custom frame escape hatch is still achievable using Alignment if you wrap your content in `.constrainedSize()`.

Reply:
Got it – this makes sense. I'll do another PR into this one to separate the changes. I plan on keeping the outer `positioning` though; and will replace corner with alignment.

Reply:
So something I don't quite understand looking at the existing alignments defined in Alignments.swift – is each of them provide a different edge – eg leading provides 0, center provides the center, and trailing provides the width – but in the case of eg the Aligned element, I need to ensure the contained wrapped element just buts up against the current edge – but this edge depends on the Alignment – but since the alignments are now just types, I can't switch over them easily... I think I'd run into the same issue here... Maybe this would be better to discuss tomorrow in 1:1? I feel like we're missing something to make this work everywhere, or I'm missing something obvious